plutus-use-cases/test/Spec/Crowdfunding.hs

```diff
@@ -62,49 +63,47 @@ scenario1 = CFScenario{..} where
         campaignCollectionDeadline = 15,
         campaignOwner              = PubKey 1
         }
-    cfWallets = Wallet <$> [1..3]
+    cfWallets = [w1, w2, w3]
     cfInitialBalances = Map.fromList [
         (PubKey 1, startingBalance),
         (PubKey 2, startingBalance),
         (PubKey 3, startingBalance)]
 
+w1, w2, w3 :: Wallet
+w1 = Wallet 1
+w2 = Wallet 2
+w3 = Wallet 3
 
--- | Generate a transaction that contributes some funds to a campaign.
+-- | Generate a transaction that contributes 600 ada to a campaign.
 --   NOTE: This doesn't actually run the validation script. The script
 --         will be run when the funds are retrieved
 makeContribution :: Property
 makeContribution = checkCFTrace scenario1 $ do
     let w = Wallet 2
-        contribution = 600
-        rest = startingBalance - fromIntegral contribution
-    processPending >>= walletNotifyBlock w
-    contrib2 (cfCampaign scenario1) contribution
-    processPending >>= walletNotifyBlock w
+        rest = startingBalance - 600
+    contrib2 600
+    processPending >>= notifyBlock
     assertOwnFundsEq w rest
 
 -- | Run a campaign with two contributions where the campaign owner collects
 --   the funds at the end
 successfulCampaign :: Property
 successfulCampaign = checkCFTrace scenario1 $ do
-    let CFScenario c [w1, w2, w3] _ = scenario1
-        updateAll' = updateAll scenario1
-    updateAll'
+    collect w1
```

Comment:
Ah, the collection trigger only checks whether there is *some* money, not whether there is enough to be allowed to collect.

There's definitely a subtlety there - the version of the collection trigger that only submits the transaction when there's enough money is fiddly because of the fund visibility issue.

Reply:
> (I'd like to see a test demonstrating that it does do what I just said if indeed it does)

That's what it should do, I'll add the test

> the collection trigger only checks whether there is some money

Do you mean the refund trigger? The collection trigger checks that the funds at the address are above the target: https://github.com/input-output-hk/plutus/pull/304/files#diff-b906bee92c907b7d2ef9c352fc08cce3R187

> the version of the collection trigger that only submits the transaction when there's enough money is fiddly because of the fund visibility issue

But the transaction that collects the funds needs to specify all the inputs explicitly, just an address is not enough. So if there are any contributions that your wallet hasn't seen yet then you can't collect them (and they can be reclaimed after the deadline).

Reply:
Doh, I was looking at the wrong trigger!

> But the transaction that collects the funds needs to specify all the inputs explicitly, just an address is not enough. So if there are any contributions that your wallet hasn't seen yet then you can't collect them (and they can be reclaimed after the deadline).

Ah, interesting. So really the only time you should set up the collection trigger is at the moment of campaign creation. So I think we should put it in there. And then the tests for late collection make less sense. But we could make them make sense using our notification technique: we can emulate "late collection" by having the creator not be notified until the deadline has passed, which will result in late collection.

Reply:
> we can emulate "late collection" by having the creator not be notified until the deadline has passed, which will result in late collection

That's implemented in the `cantCollectLate` test. I also added one where the campaign owner only sees one of the contributions and thinks that the target hasn't been reached.